sotodlib/wiregrid/calibrate.py

```diff
@@ -0,0 +1,404 @@
+#!/usr/bin/python3
+
+# import modules
+# python standard
+import os
+import sys
+import scipy
+from scipy import odr
+import numpy as np
+from iminuit import cost, Minuit
+from matplotlib.patches import Ellipse, Circle
+
+# SO specific
+from sotodlib import core
+from sotodlib.core import Context
+from sotodlib.io import hk_utils
+import sotodlib.io.load_smurf as ls
+from sotodlib.io.load_smurf import Observations
+
+wg_degpercount = 360/52000
+stream_ids = ['ufm_mv19', 'ufm_mv18', 'ufm_mv22', 'ufm_mv29', 'ufm_mv7', 'ufm_mv9', 'ufm_mv15'] # e.g. MF1
+
+def get_det_angle(aman, **kwargs):
+    """
+    Calibrate TODs and get det_angle_raw by using the wire grid.
+    Before using this code, your axis manager must have two field implemented by HWP, i.e., demodQ and demodU
+
+    Parameters:
+    - aman : AxisManager
+    - wg_config : yaml file for the settings to load the wire grid meta data
+        e.g. wg_config = '/so/home/hnakata/scratch/metadata/context/2401_wg_satp1.yaml'
+    - wg_offset : the offset angle of wires defined by the hardware of the Grid Loader
+    - stopped_time : the stopped time of your target operation
+    Return:
+    - aman : wraped with the calibration data of the wire grid. Fields are explained in each method.
+    """
+    _config = kwargs['wg_config']
+    _offset = kwargs['wg_offset']
+    _stopped_time = kwargs['stopped_time']
+    #
+    # wrap the house-keeping data of the wire grid
+    aman = wg_wrap_hk(aman, _config)
+    #
+    # restrict AxisManger within the operation range
+    idx_wg_inside = _get_operation_range(aman)
+    aman = aman.restrict('samps', (idx_wg_inside[0], idx_wg_inside[-1]), in_place=False)
+    #
+    # wrap the demodQ/demodU related with the steps of the wire grid
+    # AxisManager here must have demodQ/demodU by the HWP
```

Comment:
This function should probably have an `demodQ` and `demodU` as arguments. Typical behavior (as implemented in other modules) would be to have those arguments default set to None, then logic could be `if None` check if `demodQ` and `demodU` are in `aman` and if not either fail out or call the `hwp.demod` function.

Reply:
Yeah, this function requires that `demodQ` and `demodU` are included as attributes of aman. I have to implement a single `if` section on it.